rero_ils/facets.py

```diff
@@ -51,4 +55,146 @@ def i18n_facets_factory(search, index):
+                    elif 'date_histogram' in facet_body:
+                        facet_field = facet_body.get('date_histogram')['field']
+                    filters, filters_group, urlkwargs = \
+                        _create_filter_dsl(urlkwargs,
+                                           facets.get('post_filters', {}))
+                    facet_filter = _facet_filter(
+                                    index, filters, filters_group,
+                                    facet_name, facet_field)
+                    if facet_filter:
+                        facet_body = dict(aggs=dict(aggs_facet=facet_body))
+                        facet_body['filter'] = facet_filter
+                    aggs.update({facet_name: facet_body})
+            search = _aggregations(search, aggs)
+
+        # Query filter
+        search, urlkwargs = _query_filter(
+            search, urlkwargs, facets.get("filters", {}))
+
+        # Post filter
+        search, urlkwargs = _post_filter(
+            search, urlkwargs, facets.get("post_filters", {}))
+
+    return (search, urlkwargs)
+
+
+def _create_filter_dsl(urlkwargs, definitions):
+    """Create a filter DSL expression."""
+    filters = []
+    filters_group = {}
+    for name, filter_factory in definitions.items():
+        if isinstance(filter_factory, dict):
+            filters_group[name] = []
+            for f_name, f_filter_factory in filter_factory.items():
+                values = request.values.getlist(f_name, type=text_type)
+                if values:
+                    filters_group[name].append(f_filter_factory(values))
+                    for v in values:
+                        urlkwargs.add(f_name, v)
+        else:
+            values = request.values.getlist(name, type=text_type)
+            if values:
+                filters.append(filter_factory(values))
+                for v in values:
+                    urlkwargs.add(name, v)
+
+    return (filters, filters_group, urlkwargs)
+
+
+def _post_filter(search, urlkwargs, definitions):
+    """Ingest post filter in query."""
+    filters, filters_group, urlkwargs = \
+        _create_filter_dsl(urlkwargs, definitions)
+
+    for filter_ in filters:
+        search = search.post_filter(filter_)
+
+    for name_group, filter_ in filters_group.items():
+        q = Q('bool', should=filter_)
+        search = search.post_filter(q)
+
+    return (search, urlkwargs)
+
+
+def _facet_filter(index, filters, filters_group, facet_name, facet_field):
+    """Ingest filter in facet."""
+    s = Search(index=index)
+
+    for filter_ in filters:
+        for k, v in filter_.to_dict().items():
+            filter_field = list(v.keys())[0]
+            if filter_field != facet_field:
+                s = s.query(filter_)
+
+    for name_group, filter_ in filters_group.items():
+        if facet_name != name_group:
+            q = Q('bool', should=filter_)
+            s = s.query(q)
+
+    if facet_name == 'subject_fiction':
```

Comment:
I dislike this approach: hard coding a facet filter name here:  is it possible to create a facet filter here:
https://github.com/rero/rero-ils/pull/2953/files#diff-6be28a31924cc25a9c440ab46eedfa749b3b94ec7a6c5bb3021f76530ac95e4aL1959

Reply:
Because of the OR filters defined in post_filters in the config.py file, each facet needs to be filtered to reflect the selection made by the user on other facets. 
These filters are dynamically created by the function [_facet_filter](https://github.com/rero/rero-ils/blob/98dd50a29525a4826b12cdf7fb3b5325eeca5df4/rero_ils/facets.py#L171) in the file facets.py and a nested aggregation is created in the function [default_facets_factory](https://github.com/rero/rero-ils/blob/98dd50a29525a4826b12cdf7fb3b5325eeca5df4/rero_ils/facets.py#L117) before applying the filter.

The facets subject_fiction and subject_no_fiction have a filter independently of the other facets selection. These filters are added to the dynamically generated filters in the function _facet_filter.